From 59486b2a4ea51afb57182af31a64542fc9ff4974 Mon Sep 17 00:00:00 2001
From: Tomas Henzl <thenzl@redhat.com>
Date: Thu, 18 Sep 2014 14:14:12 -0400
Subject: [scsi] megaraid: Fail resume if MSI-X re-initialization failed

Message-id: <1411049668-28841-13-git-send-email-thenzl@redhat.com>
Patchwork-id: 96169
O-Subject: [RHEL7.1 e-stor PATCH 12/28] megaraid: Fail resume if MSI-X re-initialization failed
Bugzilla: 1088523
RH-Acked-by: Ewan Milne <emilne@redhat.com>
RH-Acked-by: Rob Evers <revers@redhat.com>

Currently the driver fails to analize MSI-X re-enablement
status on resuming and always assumes the success. This
update checks the MSI-X initialization result and fails
to resume if MSI-Xs re-enablement failed.

Signed-off-by: Alexander Gordeev <agordeev@redhat.com>
Acked-by: Kashyap Desai <Kashyap.desai@avagotech.com>
Cc: Neela Syam Kolli <megaraidlinux@lsi.com>
Cc: linux-scsi@vger.kernel.org
Cc: linux-pci@vger.kernel.org
Signed-off-by: Christoph Hellwig <hch@lst.de>
(cherry picked from commit 98c9df4942dece33ec213e94ff654cd8fa15a403)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/megaraid/megaraid_sas_base.c b/drivers/scsi/megaraid/megaraid_sas_base.c
index cc56c5c..ea03b46 100644
--- a/drivers/scsi/megaraid/megaraid_sas_base.c
+++ b/drivers/scsi/megaraid/megaraid_sas_base.c
@@ -5135,9 +5135,10 @@ megasas_resume(struct pci_dev *pdev)
   goto fail_ready_state;
 
  /* Now re-enable MSI-X */
- if (instance->msix_vectors)
-  pci_enable_msix(instance->pdev, instance->msixentry,
-    instance->msix_vectors);
+ if (instance->msix_vectors &&
+     pci_enable_msix(instance->pdev, instance->msixentry,
+       instance->msix_vectors))
+  goto fail_reenable_msix;
 
  switch (instance->pdev->device) {
  case PCI_DEVICE_ID_LSI_FUSION:
@@ -5246,6 +5247,7 @@ fail_init_mfi:
 
 fail_set_dma_mask:
 fail_ready_state:
+fail_reenable_msix:
 
  pci_disable_device(pdev);
 
-- 
1.7.1